File: cpp/0797.0_All_Paths_From_Source_to_Target.cpp
/*
Backtracking

Runtime: 20 ms, faster than 51.98% of C++ online submissions for All Paths From Source to Target.
Memory Usage: 15.6 MB, less than 31.58% of C++ online submissions for All Paths From Source to Target.
*/
class Solution {
public:
    vector<vector<int>> paths;
    
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int n = graph.size();       
        vector<int> p(1, 0);
        backtracking(n, graph, 0, p);
        return paths;
    }
    
    void backtracking(int n, vector<vector<int>>& graph, int node, vector<int> p) {
        if (node == n - 1) {
            paths.push_back(p);
            return;
        }
        for (auto child: graph[node]) {
            p.push_back(child);
            backtracking(n, graph, child, p);
            p.pop_back();
        }
    }
};

/*
to use vector<int>& p, instead of vector<int> p

Runtime: 12 ms, faster than 83.33% of C++ online submissions for All Paths From Source to Target.
Memory Usage: 11.9 MB, less than 63.83% of C++ online submissions for All Paths From Source to Target.
*/
class Solution {
public:
    vector<vector<int>> paths;
    
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int n = graph.size();       
        vector<int> p(1, 0);
        backtracking(n, graph, 0, p);
        return paths;
    }
    
    void backtracking(int n, vector<vector<int>>& graph, int node, vector<int>& p) {
        if (node == n - 1) {
            paths.push_back(p);
            return;
        }
        for (auto child: graph[node]) {
            p.push_back(child);
            backtracking(n, graph, child, p);
            p.pop_back();
        }
    }
};



/*
Runtime: 12 ms, faster than 83.33% of C++ online submissions for All Paths From Source to Target.
Memory Usage: 10.6 MB, less than 85.55% of C++ online submissions for All Paths From Source to Target.
*/
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<vector<int>> paths;
        vector<int> p(1, 0);
        backtracking(n, graph, paths, 0, p);
        return paths;
    }
    
    void backtracking(int n, vector<vector<int>>& graph, vector<vector<int>>& paths, int node, vector<int>& p) {
        if (node == n - 1) {
            paths.push_back(p);
            return;
        }
        for (auto child: graph[node]) {
            p.push_back(child);
            backtracking(n, graph, paths, child, p);
            p.pop_back();
        }
    }
};

File: problems/0797.0_All_Paths_From_Source_to_Target.py
'''
BFS

Runtime: 96 ms, faster than 89.26% of Python3 online submissions for All Paths From Source to Target.
Memory Usage: 16 MB, less than 8.23% of Python3 online submissions for All Paths From Source to Target.
'''
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        paths = []
        q = deque([(0, [0])])
        while q:
            node, p = q.popleft()
            if node == n - 1:
                paths.append(p)
            for child in graph[node]:
                q.append((child, p + [child]))
        return paths

'''
DFS

Runtime: 100 ms, faster than 74.93% of Python3 online submissions for All Paths From Source to Target.
Memory Usage: 15.5 MB, less than 76.20% of Python3 online submissions for All Paths From Source to Target.
'''

class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        paths = []
        n = len(graph)
        
        def backtracking(node, p):
            if node == n - 1:
                paths.append(p)
                return 
            for child in graph[node]:
                backtracking(child, p + [child])
        
        backtracking(0, [0])
        return paths

'''
BFS

Runtime: 113 ms, faster than 70.89% of Python3 online submissions for All Paths From Source to Target.
Memory Usage: 15.8 MB, less than 43.50% of Python3 online submissions for All Paths From Source to Target.
'''
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        paths = deque([[0]])
        ans = []
        while paths:
            p = paths.popleft()
            if p[-1] == n - 1:
                ans.append(p)
                continue
            for child in graph[p[-1]]:
                paths.append(p + [child])
        return ans 
